auth_saml/models/res_users.py

```diff
@@ -0,0 +1,243 @@
+                                else:
+                                    key = (name, lformat)
+                            else:
+                                key = name
+                        attrs[key] = list()
+                        for value in attribute.attributeValue:
+                            content = [a.exportToXml() for a in value.any]
+                            content = ''.join(content)
+                            attrs[key].append(content.decode('utf8'))
+                    except Exception as e:
+                        message = 'sso_after_response: value of an \
+                            attribute failed to decode as ascii: %s due to %s'
+                        _logger.warning(message % (attribute.dump(), str(e)))
+
+        matching_value = None
+        for k in attrs:
+            if isinstance(k, tuple) and k[0] == matching_attribute:
+                matching_value = attrs[k][0]
+                break
+
+        if not matching_value and matching_attribute == "subject.nameId":
+            matching_value = login.assertion.subject.nameId.content
+
+        elif not matching_value and matching_attribute != "subject.nameId":
+            raise Exception(
+                "Matching attribute %s not found in user attrs: %s" % (
+                    matching_attribute, attrs))
+
+        validation = {'user_id': matching_value}
+        return validation
+
+    @api.multi
+    def _auth_saml_signin(self, provider, validation, saml_response):
+        """ retrieve and sign into openerp the user corresponding to provider
+        and validated access token
+
+            :param provider: saml provider id (int)
+            :param validation: result of validation of access token (dict)
+            :param saml_response: saml parameters response from the IDP
+            :return: user login (str)
+            :raise: openerp.exceptions.AccessDenied if signin failed
+
+            This method can be overridden to add alternative signin methods.
+        """
+        token_osv = self.env['auth_saml.token']
+        saml_uid = validation['user_id']
+
+        user_ids = self.search(
+            [('saml_uid', '=', saml_uid), ('saml_provider_id', '=', provider)])
+
+        if not user_ids:
+            raise AccessDenied()
+
+        # TODO replace assert by proper raise... asserts do not execute in
+        # production code...
+        assert len(user_ids) == 1
+        user = user_ids[0]
+
+        # now find if a token for this user/provider already exists
+        token_ids = token_osv.search(
+            [('saml_provider_id', '=', provider), ('user_id', '=', user.id)])
+
+        if token_ids:
+            token_ids.write({'saml_access_token': saml_response})
+        else:
+            token_osv.create({'saml_access_token': saml_response,
+                              'saml_provider_id': provider,
+                              'user_id': user.id})
```

Comment:
Is it possible for a user to have different tokens? For example if (s)he logs in in different devices...

If that's the case, then tokens should be always created, not written, and maybe they should have an expiration date, and a cron chould delete them after that.